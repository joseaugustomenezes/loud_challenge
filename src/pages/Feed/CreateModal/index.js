import React from "react";
import Modal from "react-bootstrap/Modal";
import { Formik } from "formik";
import Form from "react-bootstrap/Form";
import { useDispatch, useSelector } from "react-redux";

import LoadingButton from "../../../components/LoadingButton";
import { createOpinion } from "../../../store/actions";

const CreateModal = ({ visible, onClose }) => {
  const dispatch = useDispatch();
  const opinion = useSelector((state) => state.createOpinion);

  const MyInnerForm = ({
    values,
    handleSubmit,
    handleChange,
    handleBlur,
    touched,
    errors,
  }) => (
    <Form onSubmit={handleSubmit}>
      <Form.Group>
        <Form.Label>Título</Form.Label>
        <Form.Control
          type="text"
          name="title"
          id="title"
          placeholder="Título"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.username}
          isInvalid={touched.title && errors.title}
        />
        <Form.Control.Feedback type="invalid">
          {errors.title}
        </Form.Control.Feedback>
      </Form.Group>
      <Form.Group>
        <Form.Label>Descrição</Form.Label>
        <Form.Control
          as="textarea"
          type="text"
          name="content"
          ogin
          id="content"
          placeholder="Descrição"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.content}
          isInvalid={touched.content && errors.content}
        />
        <Form.Control.Feedback type="invalid">
          {errors.content}
        </Form.Control.Feedback>
      </Form.Group>
      <div style={{ textAlign: "end" }}>
        <LoadingButton loading={opinion?.loading} type="submit">
          Criar
        </LoadingButton>
      </div>
    </Form>
  );

  const createOpinionForm = () => {
    return (
      <Formik
        initialValues={{ title: "", content: "" }}
        validate={(values) => {
          const errors = {};
          if (!values.title) errors.title = "Campo obrigatório";
          if (!values.content) errors.content = "Campo obrigatório";
          return errors;
        }}
        onSubmit={(values) => {
          dispatch(createOpinion(values));
        }}
      >
        {(props) => <MyInnerForm {...props} />}
      </Formik>
    );
  };

  return (
    <Modal show={visible} onHide={() => onClose()}>
      <Modal.Header closeButton>
        <Modal.Title>Nova opinião</Modal.Title>
      </Modal.Header>
      <Modal.Body>{createOpinionForm()}</Modal.Body>
    </Modal>
  );
};

export default CreateModal;
