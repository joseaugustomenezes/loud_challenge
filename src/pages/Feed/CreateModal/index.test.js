import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import CreateModal from "./index";

describe("<CreateModal />", () => {
  it("should render without errors", () => {
    const mockStore = configureStore();

    const component = mount(
      <Provider store={mockStore({})}>
        <CreateModal />
      </Provider>
    );
    expect(component).toMatchSnapshot();
    component.unmount();
  });

  it("should be visible when prop visible is set to true", () => {
    const mockStore = configureStore();

    const component = mount(
      <Provider store={mockStore({})}>
        <CreateModal visible={true} />
      </Provider>
    );
    expect(component.find("Modal").at(0).prop("show")).toBeTruthy();
    component.unmount();
  });

  it("should call onClose", () => {
    const mockFn = jest.fn();
    const mockStore = configureStore();
    const component = mount(
      <Provider store={mockStore({})}>
        <CreateModal visible={true} onClose={() => mockFn()} />
      </Provider>
    );
    component.find(".close").simulate("click");
    expect(mockFn).toHaveBeenCalled();
    component.unmount();
  });

  it("should validate empty inputs", async () => {
    const mockStore = configureStore();
    const component = mount(
      <Provider store={mockStore({})}>
        <CreateModal visible={true} />
      </Provider>
    );
    await component.find('MyInnerForm').props().submitForm();
    expect(component.find('MyInnerForm').update().find('.invalid-feedback').at(0).contains('Campo obrigatório')).toBeTruthy();
    expect(component.find('MyInnerForm').update().find('.invalid-feedback').at(1).contains('Campo obrigatório')).toBeTruthy();
    component.unmount();
  })
});
