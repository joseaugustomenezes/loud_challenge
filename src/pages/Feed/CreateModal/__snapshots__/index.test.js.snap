// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateModal /> should render without errors 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CreateModal>
    <Modal
      animation={true}
      autoFocus={true}
      backdrop={true}
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "ModalDialog",
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={true}
      onHide={[Function]}
      restoreFocus={true}
      show={false}
    >
      <Modal
        autoFocus={true}
        backdrop={true}
        backdropTransition={[Function]}
        containerClassName="modal-open"
        enforceFocus={true}
        keyboard={true}
        manager={
          BootstrapModalManager {
            "containers": Array [],
            "data": Array [],
            "handleContainerOverflow": true,
            "hideSiblingNodes": true,
            "modals": Array [],
            "scrollbarSize": 0,
          }
        }
        onEnter={[Function]}
        onEntering={[Function]}
        onEscapeKeyDown={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onHide={[Function]}
        renderBackdrop={[Function]}
        renderDialog={[Function]}
        restoreFocus={true}
        show={false}
        transition={[Function]}
      />
    </Modal>
  </CreateModal>
</Provider>
`;
